src/writer/agents.py

```diff
@@ -0,0 +1,383 @@
+        self.name = name
+        self.description = description
+        self.func = func
+        
+    def run(self, *args, **kwargs) -> Any:
+        return self.func(*args, **kwargs)
+
+class BaseAgent(ABC):
+    """Base class for all agents"""
+    def __init__(self, tools: Optional[List[Tool]] = None):
+        self.tools = tools or []
+        self.conversation_history = []
+        
+    @abstractmethod
+    def run(self, prompt: str) -> AgentOutput:
+        """Run the agent with the given prompt"""
+        pass
+
+    def add_tool(self, tool: Tool):
+        """Add a tool to the agent's toolset"""
+        self.tools.append(tool)
+
+    def _format_tool_descriptions(self) -> str:
+        """Format tool descriptions for prompt"""
+        if not self.tools:
+            return ""
+        return "Available tools:\n" + "\n".join(
+            f"- {tool.name}: {tool.description}" 
+            for tool in self.tools
+        )
+
+class ReActAgent(BaseAgent):
+    """
+    ReAct (Reasoning + Acting) agent that follows a structured thought process:
+    1. Reasoning about the current situation
+    2. Planning actions based on reasoning
+    3. Executing actions and observing results
+    4. Repeating until task completion
+    """
+    
+    def run(self, prompt: str) -> AgentOutput:
+        thoughts = []
+        actions = []
+        
+        # Add tools to system prompt if available
+        system_prompt = f"""Follow these steps to answer the user's question:
+1. Reason: Think about what needs to be done
+2. Action: Decide what action to take based on reasoning
+3. Observation: Consider the results
+4. Response: Provide final answer based on observations
+
+{self._format_tool_descriptions()}
+"""
+        # Initialize conversation with system prompt and user query
+        messages = [
+            {"role": "system", "content": system_prompt},
+            {"role": "user", "content": prompt}
+        ]
+        
+        # Add to conversation history
+        self.conversation_history.extend([
+            {"role": "user", "content": prompt}
+        ])
+        
+        response = None
+        max_steps = 5  # Prevent infinite loops
+        
+        for _ in range(max_steps):
+            # Get next action from AI
+            action_response = self._get_llm_response(messages)
+            
+            # Extract reasoning and action
+            reasoning = self._extract_reasoning(action_response)
+            if reasoning:
+                thoughts.append(reasoning)
+            
+            action = self._extract_action(action_response)
+            if action:
+                actions.append(action)
+                # Execute tool if specified
+                if action.get("tool"):
+                    tool_result = self._execute_tool(action["tool"], action.get("args", {}))
+                    messages.append({"role": "system", "content": f"Observation: {tool_result}"})
```

Comment:
This cannot be a "system" message. Our models don't support multiple system messages shuffled in.
We can make it into an "assistant" message and merge all assistant messages at the end.